### PR TITLE
RelayTextPrinter is now non-recursive. ExpandDataflow refactored

### PR DESCRIPTION
**Pull request by d-smirnov**
RelayTextPrinter made non-recursive to allow printing larger graphs ( > ~3000 nodes). There also a small refactoring in ExpandDataflow which was made bit more generalised with "node expander" extracted as separate method.

### CODE REVIEW

File: include/tvm/relay/expr_functor.h [hunk 1]
```diff
@@ -410,72 +414,82 @@ Expr PostOrderRewrite(const Expr& expr, ExprRewriter* rewriter);
...
-      fpush_to_stack(op->tuple);
-    } else {
-      // No need to expand the children directly run visit.
-      fvisit_leaf(node);
-      stack.pop();
+    } else if (const TupleGetItemNode* op = expr.as<TupleGetItemNode>()) {
+      result.push_back(op->tuple);
     }
-  }
+    return std::move(result);
```
Comment (mbrookhart) [inline include/tvm/relay/expr_functor.h:490]: Don't move on return, it prevents the compiler from optimizing

File: include/tvm/relay/expr_functor.h [hunk 2]
```diff
@@ -410,72 +414,82 @@ Expr PostOrderRewrite(const Expr& expr, ExprRewriter* rewriter);
...
+/*!
+ * \brief A struct to keep info of traversed expr in ExpandDataflow function
+ */
+struct v_info {
+  explicit v_info(Expr node_) : node{node_} {}
+  v_info(Expr node_, bool children_expanded_)
+      : node{node_}, children_expanded{children_expanded_} {};
+  Expr node{};
+  bool children_expanded{false};
+};
```
Comment (mbrookhart) [inline include/tvm/relay/expr_functor.h:426]: I don't understand why one would add the complexity of this vs a pair.

File: include/tvm/relay/expr_functor.h [hunk 3]
```diff
@@ -410,72 +414,82 @@ Expr PostOrderRewrite(const Expr& expr, ExprRewriter* rewriter);
...
+void ExpandDataflow(Expr expr, FCheckVisited fcheck_visited, FVisitLeaf fvisit_leaf,
+                    FExpandExpr fexpand_expr) {
+  std::deque<v_info> stack;
   auto fpush_to_stack = [&fcheck_visited, &stack](const Expr& expr) {
-    // The second state of the stack indicate whether the child has been
-    // expanded in the pre-order.
-    // NOTE: function will be inlined.
     if (!fcheck_visited(expr)) {
-      stack.push({expr, false});
+      stack.push_front(std::move(v_info(expr)));
```
Comment (mbrookhart) [inline include/tvm/relay/expr_functor.h:452]: Don't move on construction, it prevents certain compiler operations

File: include/tvm/relay/expr_functor.h [hunk 4]
```diff
@@ -32,11 +32,12 @@
 #include <tvm/relay/function.h>
 #include <tvm/relay/op.h>
 
+#include <deque>
 #include <stack>
```
Comment (mbrookhart) [inline include/tvm/relay/expr_functor.h:36]: This stack is no longer used

**Review by mbrookhart — CHANGES_REQUESTED** [review]
I don't understand the ExpandDataFlow refactor. It looks like you rewrote it without actually changing the behavior or using any new APIs? Could you explain the need for the refactor? 

I'm not familiar with the RelayTextPrinter, and I've never seen stack overflows when printing models with AsText or python str, I've done models over 150k nodes. What exactly is the use case that overflowed for you? Perhaps I just haven't been calling this class. We'll need a test that hits the bad behavior on the old version

ping @jroesch for the RelayTextPrinter

File: include/tvm/relay/expr_functor.h (hunk 2 above)
Comment (d-smirnov) [reply include/tvm/relay/expr_functor.h:426]: Mainly for clarity reasons as the language allows us to have have a struct with proper filed names instead of meaningless key/value pair.

File: include/tvm/relay/expr_functor.h (hunk 1 above)
Comment (d-smirnov) [reply include/tvm/relay/expr_functor.h:490]: Elaborate, please

File: include/tvm/relay/expr_functor.h (hunk 3 above)
Comment (d-smirnov) [reply include/tvm/relay/expr_functor.h:452]: Elaborate, please. As well as why one should want these optimisations to be used.

File: include/tvm/relay/expr_functor.h (hunk 1 above)
Comment (mbrookhart) [reply include/tvm/relay/expr_functor.h:490]: https://diego.assencio.com/?index=f57f25fd5a187c70fc7f34fcf5374773

File: include/tvm/relay/expr_functor.h (hunk 3 above)
Comment (mbrookhart) [reply include/tvm/relay/expr_functor.h:452]: See above, it prevents copy elision

File: include/tvm/relay/expr_functor.h (hunk 4 above)
Comment (d-smirnov) [reply include/tvm/relay/expr_functor.h:36]: fixed

File: include/tvm/relay/expr_functor.h (hunk 3 above)
Comment (d-smirnov) [reply include/tvm/relay/expr_functor.h:452]: Agree. Thanks for noticing this

File: include/tvm/relay/expr_functor.h (hunk 1 above)
Comment (d-smirnov) [reply include/tvm/relay/expr_functor.h:490]: Agree. Thanks for noticing this

**Comment by d-smirnov** [comment]
ping @jroesch

**Comment by mbrookhart** [comment]
Is it possible to add a unit test that triggers this?

**Comment by d-smirnov** [comment]
There is a unit test here [7832](https://github.com/apache/tvm/pull/7832): `tests/cpp/relay_dismantler_test.cc`
Please add `std::cout << AsText( func, false );` at the end of `foo`

**Comment by mbrookhart** [comment]
Hi @d-smirnov, sorry, I missed your response last week. Can you add that as a test here? Maybe save the result of AsText as a String and assert it's not empty?

**Review by mbrookhart — APPROVED** [review]

**Comment by mbrookhart** [comment]
Thanks @d-smirnov
